build.rs

```diff
@@ -0,0 +1,115 @@
+use crate::SqlType::{Idempotent, Migration};
+use askama::Template;
+use std::fs::File;
+use std::io::{Error, Read, Write};
+use std::path::{Path, PathBuf};
+use std::string::String;
+use std::{env, fs};
+
+#[derive(Template)]
+#[template(path = "migration-wrapper.sql", escape = "none")]
+struct MigrationWrapper<'a> {
+    filename: &'a str,
+    version: &'a str,
+    body: String,
+}
+
+#[derive(Template)]
+#[template(path = "idempotent-wrapper.sql", escape = "none")]
+struct IdempotentWrapper<'a> {
+    filename: &'a str,
+    body: String,
+}
+
+fn main() {
+    println!("cargo:rerun-if-changed=build.rs");
```

Comment:
**Suggestion:** No need to specify this for `build.rs` or `Cargo.lock`, as any change to either of those triggers recompilation/re-run anyway (for `build..rs` that's always the case. I haven't confirmed, but I think changes to `Cargo.lock` only trigger build script recompilation if build dependencies change, as you'd expect). If you want to re-run the build script when dependencies change because of expected changes to codegen output, then its better to specify the path to the codegen outputs themselves, rather than trigger a re-run if any of our dependencies change.

I'm fairly sure you'll need to be more specific about the `migration` and `templates` folders. If you want to rebuild/rerun the build script when their contents change, then you have to specify each file individually. At least that was the case when I last investigated the semantics of rerun-if-changed.

Reply:
From https://doc.rust-lang.org/cargo/reference/build-scripts.html#rerun-if-changed:

> If the path points to a directory, it will scan the entire directory for any modifications.

And based on this:

> If the build script inherently does not need to re-run under any circumstance, then emitting cargo:rerun-if-changed=build.rs is a simple way to prevent it from being re-run (otherwise, the default if no rerun-if instructions are emitted is to scan the entire package directory for changes). Cargo automatically handles whether or not the script itself needs to be recompiled, and of course the script will be re-run after it has been recompiled. Otherwise, specifying build.rs is redundant and unnecessary.

I understand that we can either remove all `rerun-if-changed` entries and have the build script re-run on any change in the project dir, or specify only those things which should cause the build script to re-run. I would prefer to keep the `rerun-if-changed` entries, to reduce recompile time.

Reply:
Ah cool! Good to know that behavior is available for directories!

RE: build script. Since you _are_ emitting some rerun-if-changed entries, emitting one for build.rs is redundant/unnecessary as mentioned in the section you quoted. It's only beneficial to emit it if you don't otherwise have any of those instructions, and you don't want to rerun the build script when any file in the crate changes, and instead only want to rerun it when just the build script changes. If you have any `rerun-if-changed` instructions other than that, then Cargo reruns it automatically when the build script is changed/recompiled.

Reply:
done